# Evaluación de Métodos de Ordenamiento: Burbuja vs. QuickSort
## Elaborado por Gabriel Carrizales
#### 10/02/2025

## Introducción  
El ordenamiento de datos es una operación fundamental en el ámbito de la informática, utilizada en una gran variedad de aplicaciones, desde bases de datos hasta algoritmos de búsqueda. En esta práctica, se evaluarán dos de los algoritmos de ordenamiento más conocidos: el Método Burbuja y QuickSort. El propósito de esta evaluación es comparar su eficiencia mediante la medición de los tiempos de ejecución al ordenar listas de diferentes tamaños. A través de este análisis, se busca comprender mejor cómo la elección del algoritmo adecuado puede impactar el rendimiento, especialmente en entornos donde la eficiencia es crucial, como en sistemas de robótica e inteligencia artificial, donde el procesamiento rápido de grandes volúmenes de datos es esencial para el funcionamiento de los sistemas.

En esta práctica, implementamos y evaluamos dos algoritmos de ordenamiento en Python:  
- **Método Burbuja** 
- **QuickSort**  

El objetivo es comparar su rendimiento mediante la medición de tiempos de ejecución con listas de diferentes tamaños.  
Esta evaluación es importante porque el ordenamiento de datos es una tarea fundamental en el desarrollo de software, especialmente en aplicaciones donde la eficiencia es clave, como en robótica e inteligencia artificial.  


1. **Burbuja**: Es un algoritmo de ordenamiento sencillo, pero bastante ineficiente, especialmente cuando se tiene una gran cantidad de datos. En el peor de los casos, requiere que se realicen muchas comparaciones y reordenamientos, lo que hace que el tiempo de ejecución crezca de manera significativa a medida que aumenta el número de elementos a ordenar. La complejidad de este algoritmo es **O(n²)**, donde **n** es el número de elementos a ordenar.

2. **QuickSort**: Este algoritmo es mucho más eficiente en comparación con el anterior. En la mayoría de los casos, realiza el ordenamiento de forma rápida al dividir el conjunto de datos en partes más pequeñas, lo que permite ordenarlas de manera más ágil. Es mucho más adecuado para grandes volúmenes de datos, ya que, en promedio, tarda mucho menos tiempo que otros métodos. Su complejidad promedio es **O(n log n)**, lo que significa que puede manejar grandes cantidades de datos de forma mucho más eficiente que otros algoritmos de ordenamiento.

In [3]:
import time
import random

def metodo_burbuja(lista):
    """Implementación del método de ordenamiento Burbuja."""
    n = len(lista)
    for i in range(n):
        for j in range(0, n - i - 1):
            if lista[j] > lista[j + 1]:
                lista[j], lista[j + 1] = lista[j + 1], lista[j]
    return lista

def quick_sort(lista):
    """Implementación del método de ordenamiento QuickSort."""
    if len(lista) < 2:
        return lista
    else:
        pivote = lista[-1]
        menores = [x for x in lista[:-1] if x < pivote]
        mayores = [x for x in lista[:-1] if x >= pivote]
        return quick_sort(menores) + [pivote] + quick_sort(mayores)

def medir_tiempo(algoritmo, lista):
    """Mide el tiempo de ejecución de un algoritmo de ordenamiento."""
    inicio = time.time()
    resultado = algoritmo(lista.copy())  # Copia la lista para evitar modificarla
    fin = time.time()
    return fin - inicio

# Generar listas de prueba de diferentes tamaños
tamaños = [1000, 5000, 10000]
resultados = {}

for n in tamaños:
    lista_prueba = [random.randint(0, 10000) for _ in range(n)]
    tiempo_burbuja = medir_tiempo(metodo_burbuja, lista_prueba)
    tiempo_quick = medir_tiempo(quick_sort, lista_prueba)
    
    resultados[n] = {
        "Burbuja": tiempo_burbuja,
        "QuickSort": tiempo_quick
    }

# Imprimir resultados
print("Comparación de tiempos de ejecución:")
for tamaño, tiempos in resultados.items():
    print(f"\nTamaño de la lista: {tamaño}")
    print(f"Método Burbuja: {tiempos['Burbuja']:.6f} segundos")
    print(f"QuickSort: {tiempos['QuickSort']:.6f} segundos")


Comparación de tiempos de ejecución:

Tamaño de la lista: 1000
Método Burbuja: 0.247848 segundos
QuickSort: 0.002999 segundos

Tamaño de la lista: 5000
Método Burbuja: 3.880609 segundos
QuickSort: 0.018987 segundos

Tamaño de la lista: 10000
Método Burbuja: 13.599612 segundos
QuickSort: 0.045970 segundos


## Pruebas de Rendimiento  

Se generaron listas de prueba de tamaños: **1000, 5000 y 10,000 elementos**.  
Cada algoritmo fue ejecutado y se midió su tiempo de ejecución utilizando la función `time()`.  


## Resultados y Análisis  

Los tiempos de ejecución obtenidos fueron los siguientes:

| Tamaño de la Lista | Método Burbuja (seg) | QuickSort (seg) |
|--------------------|--------------------|-----------------|
| 1000 elementos    | 0.2478             | 0.0029          |
| 5000 elementos    | 3.8806             | 0.0189          |
| 10,000 elementos  | 13.5996            | 0.0459          |

### **Análisis**  
- **Burbuja es ineficiente**: A medida que aumenta el tamaño de la lista, el tiempo de ejecución crece de forma cuadrática.  
- **QuickSort es mucho más rápido**, incluso con 10,000 elementos.  
- Esto confirma que **QuickSort es la mejor opción para grandes volúmenes de datos**.  


## Conclusiones  

QuickSort es significativamente más eficiente que el método Burbuja, especialmente cuando se trabaja con listas grandes, mientras que el método Burbuja solo resulta útil para listas pequeñas o con fines educativos. Este análisis destaca la importancia de elegir el algoritmo adecuado según el tamaño y la complejidad del problema, asegurando un rendimiento óptimo. Además, el uso de herramientas como Jupyter Notebook y GitHub se vuelve fundamental para documentar, organizar y compartir código de manera profesional, facilitando el trabajo colaborativo y la revisión de proyectos.